flybywiresim/a32nx
# fix(tcas): TCAS critical hotfixes and improvements

Fixes #5089

## Summary of Changes

* Fixed issue with ATC XPDR knob having missing tooltip.
* Fixed issue with ATC RPTG not triggering TCAS STBY alert and not disabling TCAS reporting
* Fixed plane spawn state on runway: ATC RPTG should be ON not OFF
* Linked green TCAS STBY alert behaviour with TCAS sensitivity level (thanks Beheh)
* Improved TCAS sensitivity level calculation logic
* Minor refactor/streamlining of TCAS computer logic
* Added more documentation/code comments

## Screenshots (if necessary)



## References

See #5089 and #3378




## Additional context



Discord username (if different from GitHub): 2Cas#1022

## Testing instructions


* Spawn in C&D + normal engine start procedure (a major airport would be good for convenience for traffic)
* Turn on EFB, go to settings > Sim Options, TCAS debug option should have been removed and be no longer present. 
* Ensure that ATC XPDR selection knob has a proper tool-tip description.
* You can either inject traffic in via ATC network, a traffic add-on or use MSFS live traffic (not AI Offline traffic). 
* Take note of TCAS STBY green alert on the right side of the Upper ECAM. This should now correctly toggle on when ALT RPTG and TCAS STBY switches are set to OFF and STBY. (TA ONLY should not appear on the ND when green TCAS STBY is displayed and TCAS is inactive)
* Takeoff with either ALT RPTG switch at OFF position, or the TCAS switch in the STBY position (not TA/RA or TA). After wheels up takeoff, the alert should turn into an amber caution (NAV TCAS STBY and TCAS STBY in amber) on the upper ECAM.
* Switch ALT RPTG to ON and TCAS to TA/RA. The caution should disappear and intruder/traffic/TCAS functionality should be restored on the ND.
* Use the dev mode Windows > Behavior debug window to modify the A32NX_SWITCH_ATC_ALT LVar to manipulate the ALT RPTG knob, see:

https://user-images.githubusercontent.com/15316958/155104546-55efd558-1c84-4c99-bd41-a87280c7aa50.mp4

* TA/RAs should be issued where appropriate, traffic should be displayed as per defined envelops depending on reporting mode (defined range: 80 nm ahead, 30nm side to side, 20 nm behind, with mode set to ABV = + 9900 ft/-2700 ft, BLW = -9900 ft/+2700ft, ALL +-2700ft. Note that MSFS Live Traffic can be buggy and show contacts that are invisible/not present/impossible, it is better to rely on traffic addons/VATSIM for this.)

* Spawn in mid-air and on the runway, ALT RPTG switch should be ON. Ensure that the TCAS STBY memo is not active on the Upper ECAM (TCAS should be functioning as normal). 


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
